### ISSUE TITLE
ValueErrors and TypeErrors

### ISSUE THREAD

**Issue by zhujik** [DontShaveTheYak/cf2tf#110]
Hello, I am trying to convert the following stack: https://datadog-cloudformation-template.s3.amazonaws.com/aws/forwarder/latest.yaml
However, I get the following errors:

```
Traceback (most recent call last):
  File "/home/zhujik/.local/bin/cf2tf", line 8, in <module>
    sys.exit(cli())
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/app.py", line 44, in cli
    config = TemplateConverter(tmpl_path.stem, cf_template, search_manger).convert()
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 94, in convert
    tf_resources = self.convert_to_tf(self.manifest)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 146, in convert_to_tf
    tf_resources.extend(converter(resources))
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 271, in convert_conditions
    resolved_values = self.resolve_values(dict_conditons, functions.ALL_FUNCTIONS)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 181, in resolve_values
    data[key] = self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 189, in resolve_values
    value = self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 197, in resolve_values
    resolved_list_values = [
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 198, in <listcomp>
    self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 187, in resolve_values
    raise ValueError(f"{key} not allowed to be nested in {prev_func}.")
ValueError: Fn::Select not allowed to be nested in Fn::Equals.
```

However, I thought that Select can actually be nested in Equals since a few years in cfn.

When I remove the only occurence of this, I get this error instead:

```
Traceback (most recent call last):
  File "/home/zhujik/.local/bin/cf2tf", line 8, in <module>
    sys.exit(cli())
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/zhujik/.local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/app.py", line 44, in cli
    config = TemplateConverter(tmpl_path.stem, cf_template, search_manger).convert()
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 94, in convert
    tf_resources = self.convert_to_tf(self.manifest)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 146, in convert_to_tf
    tf_resources.extend(converter(resources))
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 271, in convert_conditions
    resolved_values = self.resolve_values(dict_conditons, functions.ALL_FUNCTIONS)
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 181, in resolve_values
    data[key] = self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 189, in resolve_values
    value = self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 197, in resolve_values
    resolved_list_values = [
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 198, in <listcomp>
    self.resolve_values(
  File "/home/zhujik/.local/lib/python3.8/site-packages/cf2tf/convert.py", line 187, in resolve_values
    raise ValueError(f"{key} not allowed to be nested in {prev_func}.")
ValueError: Fn::Select not allowed to be nested in Fn::Equals.
```

At this point I gave up. I doubt that the cfn stack of datadog is in fact invalid though.

**Comment by shadycuz** [comment]
Thanks for reporting. Will look into this soon.

**Comment by shadycuz** [comment]
I have fixed this in #193 and it will be released soon. Here is your template, it likely has lots of issues that will need to be handled manually, but feel free to open up new issues about anything you have to fix manually.

```HCL
$ cf2tf latest.yaml 
// Converting latest.yaml to Terraform!
// Existing Terraform src code found at /tmp/terraform_src.

data "aws_caller_identity" "current" {}

data "aws_region" "current" {}

data "aws_partition" "current" {}

locals {
  mappings = {
    Constants = {
      DdForwarder = {
        Version = "3.73.0"
        LayerVersion = 30
      }
    }
  }
  IsAWSChina = data.aws_partition.current.partition == "aws-cn"
  IsGovCloud = data.aws_partition.current.partition == "aws-us-gov"
  UseZipCopier = anytrue([local.IsAWSChina, alltrue([
  var.install_as_layer == "false",
  !local.SetLayerARN
])])
  CreateDdApiKeySecret = var.dd_api_key_secret_arn == "arn:aws:secretsmanager:DEFAULT"
  SetFunctionName = !var.function_name == "DatadogForwarder"
  SetSourceZipUrl = !var.source_zip_url == ""
  SetS3SourceZip = element(split(""/"", var.source_zip_url), 0) == "s3:"
  SetDdTags = !var.dd_tags == ""
  SetDdUseTcp = var.dd_use_tcp == True
  SetDdNoSsl = var.dd_no_ssl == True
  SetDdUrl = !var.dd_url == ""
  SetDdPort = !var.dd_port == ""
  SetRedactIp = var.redact_ip == True
  SetRedactEmail = var.redact_email == True
  SetDdScrubbingRule = !var.dd_scrubbing_rule == ""
  SetDdScrubbingRuleReplacement = !var.dd_scrubbing_rule_replacement == ""
  SetExcludeAtMatch = !var.exclude_at_match == ""
  SetIncludeAtMatch = !var.include_at_match == ""
  SetDdMultilineLogRegexPattern = !var.dd_multiline_log_regex_pattern == ""
  SetDdSkipSslValidation = var.dd_skip_ssl_validation == True
  SetDdFetchLambdaTags = var.dd_fetch_lambda_tags == True
  SetDdFetchLogGroupTags = var.dd_fetch_log_group_tags == True
  SetDdFetchStepFunctionsTags = var.dd_fetch_step_functions_tags == True
  CreateS3BucketForTags = anytrue([var.dd_fetch_log_group_tags == True, var.dd_fetch_lambda_tags == True])
  SetDdUsePrivateLink = var.dd_use_private_link == True
  SetDdUseVPC = var.dd_use_vpc == True
  SetDdHttpProxyURL = !var.dd_http_proxy_url == ""
  SetDdNoProxy = !var.dd_no_proxy == ""
  SetLayerARN = !var.layer_arn == ""
  UseVPC = anytrue([local.SetDdUsePrivateLink, local.SetDdUseVPC])
  SetDdForwardLog = var.dd_forward_log == False
  SetDdUseCompression = var.dd_use_compression == False
  SetDdCompressionLevel = !var.dd_compression_level == 6
  SetDdMaxWorkers = !var.dd_max_workers == 20
  SetPermissionsBoundary = !var.permissions_boundary_arn == ""
  SetAdditionalTargetLambdas = !join("", var.additional_target_lambda_arns) == ""
  SetDdApiUrl = !var.dd_api_url == ""
  SetDdTraceIntakeUrl = !var.dd_trace_intake_url == ""
  SetDdForwarderBucketName = !var.dd_forwarder_bucket_name == ""
  SetReservedConcurrentExecutions = !var.reserved_concurrency == ""
}

variable dd_api_key {
  description = "The Datadog API key, which can be found from the APIs page (/account/settings#api). It will be stored in AWS Secrets Manager securely. If DdApiKeySecretArn is also set, this value is ignored."
  type = string
}

variable dd_api_key_secret_arn {
  description = "The ARN of the secret storing the Datadog API key, if you already have it stored in Secrets Manager. You must store the secret as a plaintext, rather than a key-value pair."
  type = string
  default = "arn:aws:secretsmanager:DEFAULT"
}

variable dd_site {
  description = "Define your Datadog Site to send data to. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com` and `ddog-gov.com`."
  type = string
  default = "datadoghq.com"
}

variable function_name {
  description = "The Datadog Forwarder Lambda function name. DO NOT change when updating an existing CloudFormation stack, otherwise the current forwarder function will be replaced and all the triggers will be lost."
  type = string
  default = "DatadogForwarder"
}

variable memory_size {
  description = "Memory size for the Datadog Forwarder Lambda function"
  type = string
  default = 1024
}

variable timeout {
  description = "Timeout for the Datadog Forwarder Lambda function"
  type = string
  default = 120
}

variable tags_cache_ttl_seconds {
  description = "TTL (in seconds) for the Datadog tags cache"
  type = string
  default = 300
}

variable reserved_concurrency {
  description = "Reserved concurrency for the Datadog Forwarder Lambda function. If empty, use unreserved account concurrency.  We recommend using at least 10 reserved concurrency, but default to 0 as you may need to increase your limits for this. If using unreserved account concurrency you may limit other lambda functions in your environment."
  type = string
}

variable log_retention_in_days {
  description = "CloudWatch log retention for logs generated by the Datadog Forwarder Lambda function"
  type = string
  default = 90
}

variable source_zip_url {
  description = "DO NOT CHANGE unless you know what you are doing. Override the default location of the function source code."
  type = string
}

variable install_as_layer {
  description = "Whether to use the layer-based installation flow. Set to false to use our legacy installation flow, which installs a second function that copies the forwarder code from Github to an S3 bucket. Defaults to true."
  type = string
  default = True
}

variable layer_arn {
  description = "ARN for the layer containing the forwarder code. If empty, the script will use the version of the layer the forwarder was published with."
  type = string
}

variable dd_tags {
  description = "Add custom tags to forwarded logs, comma-delimited string, no trailing comma, e.g., env:prod,stack:classic"
  type = string
}

variable dd_fetch_lambda_tags {
  description = "Let the forwarder fetch Lambda tags using GetResources API calls and apply them to logs, metrics and traces. If set to true, permission tag:GetResources will be automatically added to the Lambda execution IAM role. The tags are cached in memory and S3 so that they'll only be fetched when the function cold starts or when the TTL (1 hour) expires. The forwarder increments the aws.lambda.enhanced.get_resources_api_calls metric for each API call made."
  type = string
  default = True
}

variable dd_fetch_log_group_tags {
  description = "Let the forwarder fetch Log Group tags using ListTagsLogGroup and apply them to logs, metrics and traces. If set to true, permission logs:ListTagsLogGroup will be automatically added to the Lambda execution IAM role. The tags are cached in memory and S3 so that they'll only be fetched when the function cold starts or when the TTL (1 hour) expires. The forwarder increments the aws.lambda.enhanced.list_tags_log_group_api_call metric for each API call made."
  type = string
  default = True
}

variable dd_fetch_step_functions_tags {
  description = "Let the forwarder fetch Step Functions tags using GetResources API calls and apply them to logs, metrics and traces. If set to true, permission tag:GetResources will be automatically added to the Lambda execution IAM role. The tags are cached in memory and S3 so that they'll only be fetched when the function cold starts or when the TTL (1 hour) expires. The forwarder increments the aws.lambda.enhanced.get_resources_api_calls metric for each API call made."
  type = string
  default = True
}

variable dd_use_tcp {
  description = "By default, the forwarder sends logs using HTTPS through the port 443. To send logs over an SSL encrypted TCP connection, set this parameter to true."
  type = string
  default = False
}

variable dd_no_ssl {
  description = "Disable SSL when forwarding logs, set to true when forwarding logs through a proxy."
  type = string
  default = False
}

variable dd_url {
  description = "The endpoint URL to forward the logs to, useful for forwarding logs through a proxy"
  type = string
}

variable dd_port {
  description = "The endpoint port to forward the logs to, useful for forwarding logs through a proxy"
  type = string
}

variable dd_skip_ssl_validation {
  description = "Send logs over HTTPS, while NOT validating the certificate provided by the endpoint. This will still encrypt the traffic between the forwarder and the log intake endpoint, but will not verify if the destination SSL certificate is valid."
  type = string
  default = False
}

variable redact_ip {
  description = "Replace text matching \d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3} with xxx.xxx.xxx.xxx"
  type = string
  default = False
}

variable redact_email {
  description = "Replace text matching [a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+ with xxxxx@xxxxx.com"
  type = string
  default = False
}

variable dd_scrubbing_rule {
  description = "Replace text matching the supplied regular expression with xxxxx (default) or DdScrubbingRuleReplacement (if supplied). Log scrubbing rule is applied to the full JSON-formatted log, including any metadata that is automatically added by the Lambda function. Each instance of a pattern match is replaced until no more matches are found in each log. Note, using inefficient regular expression, such as `.*`, may slow down the Lambda function."
  type = string
}

variable dd_scrubbing_rule_replacement {
  description = "Replace text matching DdScrubbingRule with the supplied text"
  type = string
}

variable exclude_at_match {
  description = "DO NOT send logs matching the supplied regular expression. If a log matches both the ExcludeAtMatch and IncludeAtMatch, it is excluded. Filtering rules are applied to the full JSON-formatted log, including any metadata that is automatically added by the function. Note, using inefficient regular expression, such as `.*`, may slow down the Lambda function."
  type = string
}

variable include_at_match {
  description = "Only send logs matching the supplied regular expression and not excluded by ExcludeAtMatch. Note, using inefficient regular expression, such as `.*`, may slow down the Lambda function."
  type = string
}

variable dd_multiline_log_regex_pattern {
  description = "Use the supplied regular expression to detect for a new log line for multiline logs from S3, e.g., use expression "\d{2}\/\d{2}\/\d{4}" for multiline logs beginning with pattern "11/10/2014"."
  type = string
}

variable dd_forward_log {
  description = "Set to false to disable log forwarding, while continuing to forward other observability data, such as metrics and traces from Lambda functions."
  type = string
  default = True
}

variable dd_use_compression {
  description = "Set to false to disable log compression. Only valid when sending logs over HTTP."
  type = string
  default = True
}

variable dd_use_private_link {
  description = "DEPRECATED, DO NOT CHANGE. See README.md for details. Set to true to deploy the Forwarder to a VPC and send logs, metrics, and traces via AWS PrivateLink. When set to true, must also set VPCSecurityGroupIds and VPCSubnetIds."
  type = string
  default = False
}

variable dd_use_vpc {
  description = "Set to true to deploy the Forwarder to a VPC and send logs, metrics, and traces via a proxy. When set to true, must also set VPCSecurityGroupIds and VPCSubnetIds."
  type = string
  default = False
}

variable dd_http_proxy_url {
  description = "Sets the standard web proxy environment variables HTTP_PROXY and HTTPS_PROXY. These are the url endpoints your proxy server exposes. Don't use this in combination with AWS Private Link. Make sure to also set DdSkipSslValidation to true."
  type = string
}

variable dd_no_proxy {
  description = "Sets the standard web proxy environment variable NO_PROXY. It is a comma-separated list of domain names that should be excluded from the web proxy."
  type = string
}

variable vpc_security_group_ids {
  description = "Comma separated list of VPC Security Group Ids. Used when DdUsePrivateLink or DdUseVPC is enabled."
  type = string
}

variable vpc_subnet_ids {
  description = "Comma separated list of VPC Subnet Ids. Used when DdUsePrivateLink or DdUseVPC is enabled."
  type = string
}

variable dd_compression_level {
  description = "Set the compression level from 0 (no compression) to 9 (best compression) when sending logs."
  type = string
  default = 6
}

variable dd_max_workers {
  description = "Set the max number of workers sending logs concurrently."
  type = string
  default = 20
}

variable permissions_boundary_arn {
  description = "ARN for the Permissions Boundary Policy"
  type = string
}

variable additional_target_lambda_arns {
  description = "Comma separated list of lambda ARNs that get invoked asynchronously with the same input event"
  type = string
}

variable dd_api_url {
  description = "The endpoint URL to forward the metrics to, useful for forwarding metrics through a proxy"
  type = string
}

variable dd_trace_intake_url {
  description = "The endpoint URL to forward the traces to, useful for forwarding traces through a proxy"
  type = string
}

variable dd_forwarder_bucket_name {
  description = "The name of the forwarder bucket to create. If not provided, AWS will generate a unique name."
  type = string
}

resource "aws_lambda_function" "forwarder" {
  function_name = local.SetFunctionName ? var.function_name : null
  description = "Pushes logs, metrics and traces from AWS to Datadog."
  role = aws_iam_role.forwarder_role.arn
  handler = "lambda_function.lambda_handler"
  layers = local.UseZipCopier ? [
] : [
  local.SetLayerARN ? var.layer_arn : join(":", ["arn", data.aws_partition.current.partition, "lambda", data.aws_region.current.name, local.IsGovCloud ? "002406178527" : "464622532012", "layer", "Datadog-Forwarder", local.mappings["Constants"]["DdForwarder"]["LayerVersion"]])
]
  code_signing_config_arn = local.UseZipCopier ? {
  S3Bucket = aws_s3_bucket.forwarder_bucket.id
  S3Key = "aws-dd-forwarder-${local.mappings["Constants"]["DdForwarder"]["Version"]}.zip"
} : {
  ZipFile = " "
}
  memory_size = var.memory_size
  runtime = "python3.8"
  timeout = var.timeout
  environment {
    variables = {
      DD_ENHANCED_METRICS = "false"
      DD_API_KEY_SECRET_ARN = local.CreateDdApiKeySecret ? aws_secretsmanager_secret.dd_api_key_secret[0].id : var.dd_api_key_secret_arn
      DD_S3_BUCKET_NAME = local.CreateS3BucketForTags ? aws_s3_bucket.forwarder_bucket.id : null
      DD_SITE = var.dd_site
      DD_TAGS = local.SetDdTags ? var.dd_tags : null
      DD_TAGS_CACHE_TTL_SECONDS = var.tags_cache_ttl_seconds
      DD_FETCH_LAMBDA_TAGS = local.SetDdFetchLambdaTags ? var.dd_fetch_lambda_tags : null
      DD_FETCH_LOG_GROUP_TAGS = local.SetDdFetchLogGroupTags ? var.dd_fetch_log_group_tags : null
      DD_FETCH_STEP_FUNCTIONS_TAGS = local.SetDdFetchStepFunctionsTags ? var.dd_fetch_step_functions_tags : null
      DD_USE_TCP = local.SetDdUseTcp ? var.dd_use_tcp : null
      DD_NO_SSL = local.SetDdNoSsl ? var.dd_no_ssl : null
      DD_URL = local.SetDdUrl ? var.dd_url : null
      DD_PORT = local.SetDdPort ? var.dd_port : null
      REDACT_IP = local.SetRedactIp ? var.redact_ip : null
      REDACT_EMAIL = local.SetRedactEmail ? var.redact_email : null
      DD_SCRUBBING_RULE = local.SetDdScrubbingRule ? var.dd_scrubbing_rule : null
      DD_SCRUBBING_RULE_REPLACEMENT = local.SetDdScrubbingRuleReplacement ? var.dd_scrubbing_rule_replacement : null
      EXCLUDE_AT_MATCH = local.SetExcludeAtMatch ? var.exclude_at_match : null
      INCLUDE_AT_MATCH = local.SetIncludeAtMatch ? var.include_at_match : null
      DD_MULTILINE_LOG_REGEX_PATTERN = local.SetDdMultilineLogRegexPattern ? var.dd_multiline_log_regex_pattern : null
      DD_SKIP_SSL_VALIDATION = local.SetDdSkipSslValidation ? var.dd_skip_ssl_validation : null
      DD_FORWARD_LOG = local.SetDdForwardLog ? var.dd_forward_log : null
      DD_USE_COMPRESSION = local.SetDdUseCompression ? var.dd_use_compression : null
      DD_COMPRESSION_LEVEL = local.SetDdCompressionLevel ? var.dd_compression_level : null
      DD_MAX_WORKERS = local.SetDdMaxWorkers ? var.dd_max_workers : null
      DD_USE_PRIVATE_LINK = local.SetDdUsePrivateLink ? True : False
      DD_USE_VPC = local.UseVPC ? True : False
      HTTP_PROXY = local.SetDdHttpProxyURL ? var.dd_http_proxy_url : null
      HTTPS_PROXY = local.SetDdHttpProxyURL ? var.dd_http_proxy_url : null
      NO_PROXY = local.SetDdNoProxy ? var.dd_no_proxy : null
      DD_ADDITIONAL_TARGET_LAMBDAS = local.SetAdditionalTargetLambdas ? join(",", var.additional_target_lambda_arns) : null
      DD_API_URL = local.SetDdApiUrl ? var.dd_api_url : null
      DD_TRACE_INTAKE_URL = local.SetDdTraceIntakeUrl ? var.dd_trace_intake_url : null
    }
  }
  reserved_concurrent_executions = local.SetReservedConcurrentExecutions ? var.reserved_concurrency : null
  vpc_config = local.UseVPC ? {
  SecurityGroupIds = var.vpc_security_group_ids
  SubnetIds = var.vpc_subnet_ids
} : null
  tags = {
    dd_forwarder_version = local.mappings["Constants"]["DdForwarder"]["Version"]
  }
}

resource "aws_iam_role" "forwarder_role" {
  assume_role_policy = {
    Version = "2012-10-17"
    Statement = [
      {
        Action = [
          "sts:AssumeRole"
        ]
        Effect = "Allow"
        Principal = {
          Service = [
            "lambda.amazonaws.com"
          ]
        }
      }
    ]
  }
  managed_policy_arns = [
    "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
    "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
  ]
  permissions_boundary = local.SetPermissionsBoundary ? var.permissions_boundary_arn : null
  force_detach_policies = [
    {
      PolicyName = "ForwarderRolePolicy0"
      PolicyDocument = {
        Version = "2012-10-17"
        Statement = [
          {
            Action = [
              "s3:GetObject",
              "s3:PutObject",
              "s3:DeleteObject",
              "s3:ListBucket"
            ]
            Resource = [
              join("/", [// Unable to resolve Fn::GetAtt("ForwarderBucket.Arn"), "*"])
            ]
            Effect = "Allow"
          },
          {
            Action = [
              "s3:GetObject"
            ]
            Resource = "*"
            Effect = "Allow"
          },
          {
            Action = [
              "kms:Decrypt"
            ]
            Resource = "*"
            Effect = "Allow"
          },
          {
            Action = [
              "secretsmanager:GetSecretValue"
            ]
            Resource = local.CreateDdApiKeySecret ? aws_secretsmanager_secret.dd_api_key_secret[0].id : "${var.dd_api_key_secret_arn}*"
            Effect = "Allow"
          },
          local.SetDdFetchLambdaTags ? {
  Action = [
    "tag:GetResources"
  ]
  Resource = "*"
  Effect = "Allow"
} : null,
          local.SetDdFetchLogGroupTags ? {
  Action = [
    "logs:ListTagsLogGroup"
  ]
  Resource = "*"
  Effect = "Allow"
} : null,
          local.SetDdFetchStepFunctionsTags ? {
  Action = [
    "tag:GetResources"
  ]
  Resource = "*"
  Effect = "Allow"
} : null,
          local.UseVPC ? {
  Action = [
    "ec2:CreateNetworkInterface",
    "ec2:DescribeNetworkInterfaces",
    "ec2:DeleteNetworkInterface"
  ]
  Resource = "*"
  Effect = "Allow"
} : null,
          local.SetAdditionalTargetLambdas ? {
  Action = [
    "lambda:InvokeFunction"
  ]
  Resource = var.additional_target_lambda_arns
  Effect = "Allow"
} : null
        ]
      }
    }
  ]
  tags = {
    dd_forwarder_version = local.mappings["Constants"]["DdForwarder"]["Version"]
  }
}

resource "aws_lambda_permission" "cloud_watch_logs_permission" {
  function_name = aws_lambda_function.forwarder.arn
  action = "lambda:InvokeFunction"
  principal = local.IsAWSChina ? "logs.${data.aws_region.current.name}.amazonaws.com.cn" : "logs.${data.aws_region.current.name}.amazonaws.com"
  source_account = data.aws_caller_identity.current.account_id
}

resource "aws_lambda_permission" "s3_permission" {
  function_name = aws_lambda_function.forwarder.arn
  action = "lambda:InvokeFunction"
  principal = "s3.amazonaws.com"
  source_account = data.aws_caller_identity.current.account_id
}

resource "aws_lambda_permission" "sns_permission" {
  function_name = aws_lambda_function.forwarder.arn
  action = "lambda:InvokeFunction"
  principal = "sns.amazonaws.com"
  source_account = data.aws_caller_identity.current.account_id
}

resource "aws_lambda_permission" "cloud_watch_events_permission" {
  function_name = aws_lambda_function.forwarder.arn
  action = "lambda:InvokeFunction"
  principal = local.IsAWSChina ? "events.amazonaws.com.cn" : "events.amazonaws.com"
  source_account = data.aws_caller_identity.current.account_id
}

resource "aws_iot_thing_group" "log_group" {
  name = "/aws/lambda/${aws_lambda_function.forwarder.arn}"
  // CF Property(RetentionInDays) = var.log_retention_in_days
}

resource "aws_secretsmanager_secret" "dd_api_key_secret" {
  count = locals.CreateDdApiKeySecret ? 1 : 0
  description = "Datadog API Key"
  force_overwrite_replica_secret = var.dd_api_key
}

resource "aws_s3_bucket" "forwarder_bucket" {
  bucket = {
    ServerSideEncryptionConfiguration = [
      {
        ServerSideEncryptionByDefault = {
          SSEAlgorithm = "aws:kms"
        }
      }
    ]
  }
  // CF Property(PublicAccessBlockConfiguration) = {
  //   BlockPublicAcls = True
  //   BlockPublicPolicy = True
  //   IgnorePublicAcls = True
  //   RestrictPublicBuckets = True
  // }
}

resource "aws_s3_bucket_policy" "forwarder_bucket_policy" {
  bucket = aws_s3_bucket.forwarder_bucket.id
  policy = jsonencode({
      Version = "2012-10-17"
      Statement = [
        {
          Sid = "AllowSSLRequestsOnly"
          Action = [
            "s3:*"
          ]
          Effect = "Deny"
          Resource = [
            "${aws_s3_bucket.forwarder_bucket.arn}",
            "${aws_s3_bucket.forwarder_bucket.arn}/*"
          ]
          Condition = {
            Bool = {
              aws:SecureTransport = "false"
            }
          }
          Principal = "*"
        }
      ]
    }
  )
}

resource "aws_opsworks_custom_layer" "forwarder_zip" {
  count = locals.UseZipCopier ? 1 : 0
  // CF Property(ServiceToken) = aws_lambda_function.forwarder_zip_copier.arn
  // CF Property(DestZipsBucket) = aws_s3_bucket.forwarder_bucket.id
  // CF Property(SourceZipUrl) = local.SetSourceZipUrl ? var.source_zip_url : "https://github.com/DataDog/datadog-serverless-functions/releases/download/aws-dd-forwarder-${local.mappings["Constants"]["DdForwarder"]["Version"]}/aws-dd-forwarder-${local.mappings["Constants"]["DdForwarder"]["Version"]}.zip"
}

resource "aws_lambda_function" "forwarder_zip_copier" {
  count = locals.UseZipCopier ? 1 : 0
  description = "Copies Datadog Forwarder zip to the destination S3 bucket"
  handler = "index.handler"
  runtime = "python3.8"
  timeout = 300
  code_signing_config_arn = {
    ZipFile = "import json
import logging
import threading
import boto3
import urllib.request
import os

def send_cfn_resp(event, context, response_status):
    resp_body = json.dumps({
        'Status': response_status,
        'Reason': f'See reasons in CloudWatch Logs - group: {context.log_group_name}, stream:{context.log_stream_name}',
        'PhysicalResourceId': context.log_stream_name,
        'StackId': event['StackId'],
        'RequestId': event['RequestId'],
        'LogicalResourceId': event['LogicalResourceId'],
        'Data': {}
    }).encode('utf-8')
    req = urllib.request.Request(url=event['ResponseURL'], data=resp_body, method='PUT')
    with urllib.request.urlopen(req) as f:
        logging.info(f'Sent response to CloudFormation: {f.status}, {f.reason}')
def delete_zips(bucket):
    s3 = boto3.resource('s3')
    bucket = s3.Bucket(bucket)
    bucket.objects.all().delete()
def copy_zip(source_zip_url, dest_zips_bucket):
    s3 = boto3.client('s3')
    s3_prelude = "s3://"
    filename = "aws-dd-forwarder-{}.zip".format(os.environ.get("DD_FORWARDER_VERSION"))
    if source_zip_url.startswith(s3_prelude):
        parts = source_zip_url[len(s3_prelude):].split('/')
        bucket = parts[0]
        key = '/'.join(parts[1:])
        response = s3.get_object(Bucket=bucket, Key=key)
        data = response["Body"]
        s3.upload_fileobj(data, dest_zips_bucket, filename)
    else:
        with urllib.request.urlopen(source_zip_url) as data:
            s3.upload_fileobj(data, dest_zips_bucket, filename)
def timeout(event, context):
    logging.error('Execution is about to time out, sending failure response to CloudFormation')
    send_cfn_resp(event, context, 'FAILED')
def handler(event, context):
    # make sure we send a failure to CloudFormation if the function
    # is going to timeout
    timer = threading.Timer((context.get_remaining_time_in_millis()
              / 1000.00) - 0.5, timeout, args=[event, context])
    timer.start()
    logging.info(f'Received event: {json.dumps(event)}')
    try:
        source_zip_url = event['ResourceProperties']['SourceZipUrl']
        dest_zips_bucket = event['ResourceProperties']['DestZipsBucket']
        if event['RequestType'] == 'Delete':
            delete_zips(dest_zips_bucket)
        else:
            copy_zip(source_zip_url, dest_zips_bucket)
    except Exception as e:
        logging.exception(f'Exception when copying zip from {source_zip_url} to {dest_zips_bucket}')
        send_cfn_resp(event, context, 'FAILED')
    else:
        send_cfn_resp(event, context, 'SUCCESS')
    finally:
        timer.cancel()
"
  }
  environment {
    variables = {
      DD_FORWARDER_VERSION = local.mappings["Constants"]["DdForwarder"]["Version"]
    }
  }
  role = aws_iam_role.forwarder_zip_copier_role.arn
}

resource "aws_cloudformation_stack_set" "forwarder_zip_ready" {}

resource "aws_iam_role" "forwarder_zip_copier_role" {
  count = locals.UseZipCopier ? 1 : 0
  assume_role_policy = {
    Version = "2012-10-17"
    Statement = [
      {
        Action = [
          "sts:AssumeRole"
        ]
        Effect = "Allow"
        Principal = {
          Service = [
            "lambda.amazonaws.com"
          ]
        }
      }
    ]
  }
  managed_policy_arns = [
    "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
    "arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
  ]
  permissions_boundary = local.SetPermissionsBoundary ? var.permissions_boundary_arn : null
  force_detach_policies = [
    {
      PolicyName = "ForwarderZipCopierRolePolicy0"
      PolicyDocument = {
        Version = "2012-10-17"
        Statement = [
          {
            Effect = "Allow"
            Action = [
              "s3:PutObject",
              "s3:DeleteObject"
            ]
            Resource = [
              join("/", [// Unable to resolve Fn::GetAtt("ForwarderBucket.Arn"), "*"])
            ]
          },
          {
            Effect = "Allow"
            Action = [
              "s3:ListBucket"
            ]
            Resource = [
                            // Unable to resolve Fn::GetAtt("ForwarderBucket.Arn")
            ]
          },
          local.SetS3SourceZip ? {
  Effect = "Allow"
  Action = [
    "s3:GetObject"
  ]
  Resource = [
    join("", ["arn:*:s3:::", element(split(""s3://"", var.source_zip_url), 1)])
  ]
} : null
        ]
      }
    }
  ]
}

output "datadog_forwarder_arn" {
  description = "Datadog Forwarder Lambda Function ARN"
  value = aws_lambda_function.forwarder.arn
}

output "dd_api_key_secret_arn" {
  description = "ARN of SecretsManager Secret with Datadog API Key"
  value = aws_secretsmanager_secret.dd_api_key_secret[0].id
}

output "forwarder_bucket_name" {
  description = "Name of the S3 bucket used by the Forwarder"
  value = aws_s3_bucket.forwarder_bucket.id
}
```
